contracts/optimism/TokenRateOracle.sol

```diff
@@ -240,31 +326,55 @@ contract TokenRateOracle is CrossDomainEnabled, ITokenRateOracle, Versioned {
         return false;
     }
 
-    function _setTokenRateAndTimestamps(uint128 tokenRate_, uint64 rateUpdateL1Timestamp_) internal {
-        _loadTokenRateData().value = TokenRateData(tokenRate_, rateUpdateL1Timestamp_, uint64(block.timestamp));
+    function _addTokenRate(
+        uint128 tokenRate_, uint64 rateUpdateL1Timestamp_, uint64 rateReceivedL2Timestamp_
+    ) internal {
+        _getStorageTokenRates().push(TokenRateData({
+            tokenRate: tokenRate_,
+            rateUpdateL1Timestamp: rateUpdateL1Timestamp_,
+            rateReceivedL2Timestamp: rateReceivedL2Timestamp_
+        }));
     }
 
-    function _getTokenRateAndTimestamps() private view returns (TokenRateData memory tokenRateData) {
-        tokenRateData = _loadTokenRateData().value;
+    function _getLastTokenRate() internal view returns (TokenRateData storage) {
+        if (_getStorageTokenRates().length == 0) {
```

Comment:
I think we don't need this check because an element is always added in `initialize` and there is only way to remove elements _after_ given index